File: 二分查找/34.cpp
#include <vector>
using namespace std;
class Solution
{
private:
    int binarySearch(vector<int>&nums ,int target){
        int l=0;
        int h=nums.size()-1;
        int idx = -1;
        while(l<=h){
            idx = int((l+h)/2);
            if(nums[idx]>target){
                h = idx-1;
            }
            else if(nums[idx]<target){
                l = idx+1;
            }
            else{
                return idx; 
            }
        }
        return -1;
    }
public:
    vector<int> searchRange(vector<int> &nums, int target)
    {
        int anchor = binarySearch(nums,target);
        if(anchor==-1) return {-1,-1};
        vector<int> res;
        int first = anchor;
        while((first>0)&&(nums[first-1]==target)) first--;
        res.push_back(first);
        int second = anchor;
        while(((second)<nums.size()-1)&&(nums[second+1]==target)) second++;
        res.push_back(second);
        return res;
    }
};

File: 二分查找/34.py
'''
给定一个按照升序排列的整数数组 nums，和一个目标值 target。找出给定目标值在数组中的开始位置和结束位置。

如果数组中不存在目标值 target，返回 [-1, -1]。

进阶：
你可以设计并实现时间复杂度为 O(log n) 的算法解决此问题吗？
'''


class Solution:
    def find(self, arr, target):
        l = 0
        h = len(arr)-1
        while(l <= h):
            idx = (l+h)//2
            if(arr[idx] > target):
                h = idx-1
            elif(arr[idx] < target):
                l = idx+1
            else:
                return idx
        return -1

    def searchRange(self, nums, target):
        anchor = self.find(nums, target)
        if anchor == -1:
            return [-1, -1]
        res = []
        idx = anchor

        while (idx > 0) and (nums[idx-1] == target):
            idx -= 1
        res.append(idx)

        idx = anchor

        while (idx < len(nums)-1) and (nums[idx+1] == target):
            idx += 1
        res.append(idx)
        return res


arr = [1,4]
obj = Solution()
res = obj.searchRange(arr, 4)
print(res)
